### ISSUE TITLE
slow load for configure

### ISSUE THREAD

**Issue by dmuhamedagic** [ClusterLabs/crmsh#459]
Every crm cf invocation takes almost two seconds:

```
root@alpaca1:~# time crm cf  <> /dev/null

real	0m1.709s
user	0m0.310s
sys	0m0.175s
```

It turns out that most of the time is used by `crm_resource --list-standards` (for whatever reason is that one taking so long, but that's another matter).
This is cached for the completion, but the completion is not used for one-shot commands such as `crm cf show blah`.

**Comment by dmuhamedagic** [comment]
No comments? Don't you find this an important usability issue?

**Comment by dmuhamedagic** [comment]
As for `crm_resource --list-standards`, the culprit is actually systemd and I see the same behaviour with `systemctl --list-unit-files`. It makes around 300k filesystem related system calls (for instance, readlinkat was invoked 119684 times). This is on a Debian Buster.

**Comment by gao-yan** [comment]
Hi Dejan,

It's definitely an usability issue if it's so slow. I gave it a quick check with my different releases of SUSE Linux. It looks much better though.

`crm_resource --list-standards` takes around 0.06s for SLE12SP4,  0.25s for opensuse 15.1.

`time crm cf  <> /dev/null` takes around 0.2s for SLE12 SP4, 0.45s for opensuse 15.1.

`systemctl list-unit-files >/dev/null` takes around 0.05s for SLE12 SP4,  0.25s for opensuse 15.1.

A more similar command `crm_resource --list-agents systemd` takes  0.06s for SLE12 SP4,  0.23s for opensuse15.1.

The performance is acceptable AFAICT, although there are kind of big differences between the releases. They are both VMs BTW.  I assume the results could quite depend on different systems, environments, and/or probably versions of systemd ...  --  Debian Buster seems like a very new release. Not sure if there's performance regression related to systemd though...

**Comment by dmuhamedagic** [comment]
Hey Yan,
Adding --state=enabled to the list-unit-files command makes it run twice as long (3.2s on an Intel Atom). That's really dreadful. 
With Debian Stretch (systemd 232) I get similar results as you with SLE12SP4 (0.05s), so there's definitely a regression between that version and systemd 241 of Debian Buster. I also have one system running a version in between (237) and its performance is also in between (~0.5s). Opensuse 15.1 runs version 234 shows a 5x regression in terms of performance. So, apparently the regression is cumulative :-D
Anyway, it would be interesting to see how it fairs on a Factory system. Even though the performance is fine with SLE12SP4, I guess that this problem is around the corner for SLE too.

**Comment by gao-yan** [comment]
> Hey Yan,
> Adding --state=enabled to the list-unit-files command makes it run twice as long (3.2s on an Intel Atom). That's really dreadful.

Tried it with --state=enabled, the option itself didn't show me obvious performance drop though.

> With Debian Stretch (systemd 232) I get similar results as you with SLE12SP4 (0.05s), so there's definitely a regression between that version and systemd 241 of Debian Buster. I also have one system running a version in between (237) and its performance is also in between (~0.5s). Opensuse 15.1 runs version 234 shows a 5x regression in terms of performance. So, apparently the regression is cumulative :-D

Indeed it looks like so :-) Interesting ...

> Anyway, it would be interesting to see how it fairs on a Factory system. Even though the performance is fine with SLE12SP4, I guess that this problem is around the corner for SLE too.

I tried the latest version systemd-243 in opensuse Factory. `systemctl list-unit-files` takes around 0.7s. So indeed there seems to be more performance regression. 

@Werkov, I hope you are the right person to ask :-) Do you have any clue about this? Thanks!

**Comment by dmuhamedagic** [comment]
Forgot to mention add the affected version (Debian Buster):
[0]capote:config > systemctl --version
systemd 241 (241)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid

I also opened a Debian bug report: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=946961

**Comment by Werkov** [comment]
I'm afraid I won't help here. If the regression is observed on systems with comparable number of unit files, then it needs profiling PID 1 and systemctl to find the culprit. crmsh is a wrong tree to bark up.

([1] I observe runtime of 0.97 s on my system with 531 unit files and systemd v243, it's independent of the filtered set size and also `SYSTEMCTL_INSTALL_CLIENT_SIDE=1 systemctl` (debug variable to bypass PID 1 and read unit files directly) takes similar amount of time (sys 0.55s usr 0.41s).)

**Comment by dmuhamedagic** [comment]
Yes, crmsh is a victim here, though it could be improved by being lazy and invoking external programs for the sake of completion only when it's really needed. IIRC, it used to behave like that before.
I suspect that with bypassing the systemd, systemctl exercises the very same code, and if so the result is not surprising.

**Comment by dmuhamedagic** [comment]
> > With Debian Stretch (systemd 232) I get similar results as you with SLE12SP4 (0.05s), so there's definitely a regression between that version and systemd 241 of Debian Buster. I also have one system running a version in between (237) and its performance is also in between (~0.5s). Opensuse 15.1 runs version 234 shows a 5x regression in terms of performance. So, apparently the regression is cumulative :-D
> 
> Indeed it looks like so :-) Interesting ...

Yes, and I'm somewhat curious how this came to be. Knuth once told one of his students (paraphrasing): "Please never again use linear search!" ;-)

**Comment by gao-yan** [comment]
> I'm afraid I won't help here. 

Thought you were working on systemd :-)

> If the regression is observed on systems with comparable number of unit files, then it needs profiling PID 1 and systemctl to find the culprit. crmsh is a wrong tree to bark up.

crmsh invokes `crm_resource --list-standards` command which does the request `ListUnitFiles` from systemd, which is assumably  similar to `systemctl list-unit-files`. That's the reason why crmsh suffers from the performance regression.

Although it was tested in different systems, there were comparable numbers of unit files:

- With 233 unit files, it took 0.06s with systemd-228
- With 262 unit files, it took 0.25s with systemd-234
- With 346 unit files, it took 0.7s with systemd 243

`systemctl list-unit-files` got basically the same results. It''s kind of big regression.

> 
> ([1] I observe runtime of 0.97 s on my system with 531 unit files and systemd v243, it's independent of the filtered set size and also `SYSTEMCTL_INSTALL_CLIENT_SIDE=1 systemctl` (debug variable to bypass PID 1 and read unit files directly) takes similar amount of time (sys 0.55s usr 0.41s).)

Indeed similar results. Anyway the performance regression is quite obvious compared with earlier versions.

**Comment by Werkov** [comment]
Cross-linking systemd/systemd#20328 or systemd/systemd#14730 (IOW this issue may be another reason to optimize it in systemd).
